### ISSUE TITLE
Feedback Request: Execute link per statement setting

### ISSUE THREAD

**Issue by mikeburgh** [dbcodeio/public#143]
As raised in #140 should I add a setting to enable an execute codelens (button) to appear above each executable statement in a file, as shown here:

![Image](https://github.com/user-attachments/assets/00fa5fb9-af9c-4f35-8bd3-a14c7cda4b0f)

If you would like to see this, please leave a comment on this issue.

**Comment by arc87** [comment]
Another suggestion may be to execute the SQL query from the current position of the cursor using the Ctrl+Enter combination, similar to how it works in the DBeaver application.

**Comment by devpkr1** [comment]
Currently, pressing Ctrl+Enter executes all statements in a script if no specific statement is highlighted or selected. However, it would be more efficient if Ctrl+Enter executed only the statement where the cursor is positioned, using delimiters to identify individual statements. This would streamline the workflow by eliminating the need to manually select or highlight statements.

Enabling an "Execute" button (codelens) above each executable statement might provide convenience but risks making the file visually cluttered, which could hinder readability.

**Comment by devpkr1** [comment]
> [@devpkr1](https://github.com/devpkr1)
> 
> I agree with you and thinking a little more deeply, including some button or link to execute the SQL query may have some drawbacks, for example, when having a long query, it would be a bit annoying to move (as an option) to the beginning of the query, instead executing the query with Ctrl+Enter would be more efficient and productive
> 
> [@mikeburgh](https://github.com/mikeburgh)
> 
> I discarded the option of adding independent buttons for executing queries, I kept the option of using Ctrl+Enter

Yes, that's sounds good. The `Ctrl+Enter` command is helpful, but it could be improved by executing only the statement at the current position, identified by a delimiter. However, this approach carries certain risks during DML operations. Highlighting statements every time for execution is safest way but can be tedious, especially when working with complex or large SQL files.

**Comment by mikeburgh** [comment]
Okay I have a rough proof of concept this for this working:

https://github.com/user-attachments/assets/2a256659-5c24-4586-ae9f-4e6e7df7a7f5

Some things to note: 
- It will highlight the editor background for what it thinks is the current statement, this would be what gets executed using the execute button, or the keyboard shortcut when there is no selection

- If there is a selection, the text changes to execute selection, and the both that link and the keyboard shortcut would execute the selection only.

A question:
- When you are out of a statement (24 second mark int he video) no codelens or highlighting appears, I think this is safest.. rather than leaving the last one present.. this way it's explicit.. and you cannot execute code you are not expecting to. But happy to discuss.

I think this solves a few problems: 
- The unreliable positioning of the execute/connection links (what VSCode calls codelens)
- Executing code without selecting it, but being clear about what's being executed
- Show statement boundaries without cluttering the interface

Ideally I would like to move the execute link to a gutter icon, but that's not possible at the moment, so the codelens is the best option.. it's a little jumpy as it pushes the text around when rendering, but it stays fixed now.

**Comment by devpkr1** [comment]
Hi Mike,

Your proof of concept is clear and effective!
Great work solving execution clarity and statement boundary issues!

**Comment by arc87** [comment]
Hi Mike, good job.

**Comment by mikeburgh** [comment]
Thanks, all the contortion since we don't have a toolbar! 

I just pushed 1.6.3 with this change.. It feels better, but it is different to before.<br><br>The only thing that's been lost is there is now no way to execute the whole script without selecting it all first.. Probably for the best as there was a setting to catch and warn just for that in case someone did it by accident.<br><br>I also left the keyboard shortcut working in notebooks to execute the selection out to the main results panel, didn't want to mess with the notebook ui of the play button, but figured the keyboard shortcut was a nice escape hatch if you want to kick results out to the main panel.

**Comment by devpkr1** [comment]
The updates sound like a solid improvement, even if they introduce some changes.
Overall, these tweaks prioritize clarity, safety, and user intent. Nicely done!

**Comment by Steuv1871** [comment]
Just tested it and it works flawlessly.
The "line shift" from codelens is not an issue.

**Comment by arc87** [comment]
> Thanks, all the contortion since we don't have a toolbar!
> 
> I just pushed 1.6.3 with this change.. It feels better, but it is different to before.The only thing that's been lost is there is now no way to execute the whole script without selecting it all first.. Probably for the best as there was a setting to catch and warn just for that in case someone did it by accident.I also left the keyboard shortcut working in notebooks to execute the selection out to the main results panel, didn't want to mess with the notebook ui of the play button, but figured the keyboard shortcut was a nice escape hatch if you want to kick results out to the main panel.

Hi Mike,

I have been testing the new features with the latest update, for now everything works perfectly, 

Regarding what you mentioned about not being able to execute the complete script, in my case it is not complicated, I use keyboard shortcuts a lot, I select all the text (Ctrl+A) and execute (Ctrl+Enter)

**Comment by mikeburgh** [comment]
Thanks everyone for the feedback, much appreciated!

[arc87](https://linear.app/dbcode/customer/arc87-7e8fda4b7cc9) that's a good point on select all, and is probably better as it's a specific action not an unintentional one.
